### ISSUE TITLE
There should be no unused return values for Enum.reduce()

### ISSUE THREAD

**Issue by devshane** [rrrene/credo#520]
### Environment

* Credo version (`mix credo -v`): 0.9.0
* Erlang/Elixir version (`elixir -v`): 1.6.4
* Operating system: OS X

### What were you trying to do?

```
Agent.update(__MODULE__, fn _ ->
  Enum.reduce(urls, &expire_url/2)
end)
```

### Expected outcome

No warnings.

### Actual outcome

`There should be no unused return values for Enum.reduce().`

**Comment by rrrene** [comment]
@devshane Thx for reporting this! :+1:

This should not happen. Unfortunately, I can not reproduce your error.

This test does not yield an issue for me:

```elixir
defmodule CredoTest do
  defp test do
    urls = []

    Agent.update(__MODULE__, fn _ ->
      Enum.reduce(urls, &expire_url/2)
    end)
  end

  defp expire_url(_, _), do: nil
end
```

Can you provide any more info about the context in which the issue occurs?

**Comment by devshane** [comment]
@rrrene I've since refactored this code because the approach felt wrong, but I can reproduce it with this:

```
defmodule CredoTest do
  @moduledoc """
  Documentation for CredoTest.
  """

  @doc """
  ## Examples

      iex> CredoTest.agent_update
      6
  """
  def agent_update do
    Agent.start_link(fn -> 0 end, name: __MODULE__)

    Agent.update(__MODULE__, fn _ ->
      Enum.reduce([1, 2, 3], fn a, b -> a + b end)
    end)

    Agent.get(__MODULE__, fn val -> val end)
  end
end
```

Using

```
  defp deps do
    [
      {:credo, "~> 0.9.0-rc1", only: [:dev, :test], runtime: false}
    ]
```

Warning:

```
[W] ↗ There should be no unused return values for Enum.reduce().
       lib/credo_test.ex:16:7 #(CredoTest.agent_update)
```

Thanks.

**Comment by Selvaticus** [comment]
Hi,

Just out of interest I've tried to reproduce, and:

- With @rrrene code no credo warning.
- With @devshane code I get the credo warning.

Same `credo` dependency both times, and the code was straight `copy` `paste`.

### System
- Credo: 0.9.0
- Elixir: 1.5.1
- OTP: 20
- OS: macOS High Sierra 10.13.3

### Funny thing (maybe you will understand, I didn't)

This code **has no**  `credo` warnings

```elixir
defmodule CredoTest do
  @moduledoc """
  Documentation for CredoTest.
  """

  @doc """
  ## Examples

      iex> CredoTest.agent_update
      6
  """
  def agent_update do
    Agent.start_link(fn -> 0 end, name: __MODULE__)

    Agent.update(__MODULE__, fn _ ->
      Enum.reduce([1, 2, 3], fn a, b -> a + b end)
    end)

  end
end
```

Removing this line, made all the difference:
```elixir
Agent.get(__MODULE__, fn val -> val end)
```

**Comment by devshane** [comment]
This is also happening with `Enum.reject`:

```
    Agent.update(__MODULE__, fn _ ->
      urls
      |> Enum.map(&process_url/1)
      |> Enum.reject(&is_nil/1)
    end)
```

**Comment by rrrene** [comment]
@devshane @Selvaticus This is a nasty bug. We should not go `1.0` without solving this. 

Thanks for again reporting this! :+1:

**Comment by Selvaticus** [comment]
Re-tested looks like it's sorted :+1:

**Comment by rrrene** [comment]
Thx for reporting this! :+1: It should be be fixed in the latest version of Credo.

If it is not, please feel free to re-open this issue!

**Comment by marinakr** [comment]
Hi!
I updated credo from master, but issue still occurs:
```
      fields
      |> List.delete(:strategy)
      |> Enum.sort()
      |> Enum.reduce(0, fn field, priority ->
        with {:ok, chunked_tasks} <- RedisApi.hget(job_id_key, field) do
          priority = create(chunked_tasks, job_id, strategy, priority)
          RedisApi.hdel(job_id_key, field)
          priority
        end
      end)
```
I don't care about final unused return value for Enum.reduce, but accumulator (priority) is necessary during reduce

**Comment by rrrene** [comment]
@marinakr That is a valid edge case! Thx for reporting it :+1:

Have a look at the latest [CHANGELOG](https://github.com/rrrene/credo/blob/aae5abab08d86ca7269613753cc425e12be83692/CHANGELOG.md). You can now ignore `reduce` if you want to in Credo `1.3.0-rc2`.
